relayer/src/link.rs

```diff
@@ -395,197 +393,263 @@ fn build_ack_from_recv_event(
     build_ack_packet(dst_chain, src_chain, &event, event_height)
 }
 
-/// From each sequence it builds either a MsgRecvPacket or a MsgTimeout message
-/// MsgTimeout-s are sent to the source chain while MsgRecvPacket-s are sent to
-/// the destination chain
-fn build_recv_packet_and_timeout_msgs(
-    packet_src_chain: Box<dyn ChainHandle>,
-    src_query_height: Height,
+struct PacketMsgCollector {
     packet_dst_chain: Box<dyn ChainHandle>,
-    dst_query_height: Height,
-    src_channel_id: &ChannelId,
-    src_port: &PortId,
-    sequences: Vec<Sequence>,
-) -> Result<(Vec<Any>, Vec<Any>), Error> {
-    if sequences.is_empty() {
-        return Ok((vec![], vec![]));
+    packet_src_chain: Box<dyn ChainHandle>,
+    opts: PacketEnvelope,
+    recv_seqs: Vec<Sequence>,
+    ack_seqs: Vec<Sequence>,
+    src_query_height: Height, // proof height for recv packets
+    dst_msgs: Vec<Any>,       // recv packets to be send to destination chain
+    dst_query_height: Height, // proof height for acks and timeout
+    src_msgs: Vec<Any>,       // acks and/or timeouts to be sent to source chain
+}
+
+impl PacketMsgCollector {
```

Comment:
This is a sort of a helper to do all logic related to sending packets?

Reply:
Yes. Before we had all the action functions returning the information as they were retrieving information. Now all is collected in this struct so we don't pass it around from return function values to input parameters to next ones.
Also, I was hoping that I will be able to eventually merge this and the Link struct but it's not as straightforward as it seems at first look. So we will look at this in subsequent PRs.

Reply:
Cool,. Merging them together in some coherent struct would be nice to have, but the modifications here are already pretty good!